File: BE/zone_service/src/common/health/health-heartbeat-publisher.ts
/**
 * Health Heartbeat Publisher
 * Publishes health status to Kafka at regular intervals
 */

import { kafkaService } from '../kafka/kafka.service';
import { logger } from '../logger/logger.service';
import { config } from '@config/config';
import axios from 'axios';
import * as os from 'os';

const TOPIC_HEALTH_STATUS = 'health-status';
const DEFAULT_PING_INTERVAL_SECONDS = 10;

interface HealthStatusDto {
  serviceName: string;
  status: string;
  timestamp: string;
  version: string;
  metadata: {
    hostname: string;
    pid: number;
    timestamp_millis: number;
  };
}

class HealthHeartbeatPublisher {
  private static instance: HealthHeartbeatPublisher | null = null;
  private intervalId: NodeJS.Timeout | null = null;
  private isRunning: boolean = false;
  private pingIntervalSeconds: number = DEFAULT_PING_INTERVAL_SECONDS;
  private hasError: boolean = false;

  private constructor() {}

  /**
   * Get singleton instance
   */
  public static getInstance(): HealthHeartbeatPublisher {
    if (!HealthHeartbeatPublisher.instance) {
      HealthHeartbeatPublisher.instance = new HealthHeartbeatPublisher();
    }
    return HealthHeartbeatPublisher.instance;
  }

  /**
   * Start publishing health heartbeat
   */
  public async start(): Promise<void> {
    if (this.isRunning) {
      logger.warn('[HealthHeartbeatPublisher] Already running');
      return;
    }

    logger.info('[HealthHeartbeatPublisher] Starting health heartbeat publisher');

    // Load ping interval from Settings Service
    try {
      const settingsServiceUrl = config.settings.serviceUrl || process.env.SETTINGS_SERVICE_URL || 'http://localhost:21502';
      const url = `${settingsServiceUrl}/api/v1/settings/HEALTH_PING_INTERVAL_SECONDS/value?defaultValue=${DEFAULT_PING_INTERVAL_SECONDS}`;
      
      const response = await axios.get(url, { timeout: 5000 });
      if (response.data && typeof response.data === 'string') {
        this.pingIntervalSeconds = parseInt(response.data, 10) || DEFAULT_PING_INTERVAL_SECONDS;
      }
      logger.info(`[HealthHeartbeatPublisher] Health ping interval: ${this.pingIntervalSeconds} seconds`);
    } catch (error) {
      logger.warn(`[HealthHeartbeatPublisher] Failed to load HEALTH_PING_INTERVAL_SECONDS from Settings Service, using default: ${DEFAULT_PING_INTERVAL_SECONDS}s`, { error });
    }

    // Wait for Kafka to be ready
    if (!kafkaService.getConnectionStatus()) {
      logger.warn('[HealthHeartbeatPublisher] Kafka not connected, will retry after Kafka is initialized');
      // Retry after 5 seconds
      setTimeout(() => this.start(), 5000);
      return;
    }

    this.isRunning = true;
    this.hasError = false;

    // Publish immediately
    await this.publishHealthStatus();

    // Then publish at regular intervals
    this.intervalId = setInterval(() => {
      this.publishHealthStatus().catch((error) => {
        logger.error('[HealthHeartbeatPublisher] Error in scheduled publish', { error });
      });
    }, this.pingIntervalSeconds * 1000);

    logger.info(`[HealthHeartbeatPublisher] Health heartbeat publisher started (interval: ${this.pingIntervalSeconds}s)`);
  }

  /**
   * Publish health status to Kafka
   */
  private async publishHealthStatus(): Promise<void> {
    // Skip if we've encountered a fatal error
    if (this.hasError) {
      return;
    }

    // Skip if Kafka is not connected
    if (!kafkaService.getConnectionStatus()) {
      logger.debug('[HealthHeartbeatPublisher] Kafka not connected, skipping publish');
      return;
    }

    try {
      const healthStatus: HealthStatusDto = {
        serviceName: 'zone-service',
        status: 'UP',
        timestamp: new Date().toISOString(),
        version: process.env.npm_package_version || '1.2.0',
        metadata: {
          hostname: os.hostname(),
          pid: process.pid,
          timestamp_millis: Date.now(),
        },
      };

      // Send with serviceName as key (required by API Gateway consumer)
      await kafkaService.sendMessage(TOPIC_HEALTH_STATUS, healthStatus, 'zone-service');
      logger.debug('[HealthHeartbeatPublisher] Health status published successfully');
    } catch (error: any) {
      logger.error('[HealthHeartbeatPublisher] Failed to publish health status', { error });

      // Check if error is due to topic not existing
      if (error?.message?.includes('UnknownTopicOrPartition') || error?.message?.includes('Topic does not exist')) {
        logger.error('[HealthHeartbeatPublisher] Topic does not exist. Stopping heartbeat.');
        this.hasError = true;
        this.stop();
      }
    }
  }

  /**
   * Stop publishing health heartbeat
   */
  public stop(): void {
    if (this.intervalId) {
      clearInterval(this.intervalId);
      this.intervalId = null;
    }
    this.isRunning = false;
    logger.info('[HealthHeartbeatPublisher] Health heartbeat publisher stopped');
  }

  /**
   * Check if publisher is running
   */
  public isActive(): boolean {
    return this.isRunning && !this.hasError;
  }
}

// Export singleton instance
export const healthHeartbeatPublisher = HealthHeartbeatPublisher.getInstance();

// Export class for advanced use cases
export { HealthHeartbeatPublisher };

File: BE/Settings_service/src/main/java/com/ds/setting/infrastructure/health/HealthHeartbeatPublisher.java
package com.ds.setting.infrastructure.health;

import com.ds.setting.business.v1.services.SettingsService;
import com.fasterxml.jackson.annotation.JsonFormat;
import lombok.Builder;
import lombok.Data;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.boot.context.event.ApplicationReadyEvent;
import org.springframework.context.event.EventListener;
import org.springframework.kafka.core.KafkaTemplate;
import org.springframework.kafka.support.SendResult;
import org.springframework.scheduling.annotation.Async;
import org.springframework.stereotype.Component;

import jakarta.annotation.PreDestroy;
import java.time.LocalDateTime;
import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicBoolean;

/**
 * Singleton component to publish health status heartbeat to Kafka
 * Publishes health status at regular intervals (configurable via Settings Service)
 */
@Slf4j
@Component
@RequiredArgsConstructor
public class HealthHeartbeatPublisher {

    private static final String TOPIC_HEALTH_STATUS = "health-status";
    
    @Value("${spring.application.name:settings-service}")
    private String serviceName;
    
    @Value("${info.app.version:1.2.0}")
    private String serviceVersion;
    
    private final KafkaTemplate<String, Object> kafkaTemplate;
    private final SettingsService settingsService;
    
    private ScheduledExecutorService scheduler;
    private final AtomicBoolean isRunning = new AtomicBoolean(false);
    private final AtomicBoolean hasError = new AtomicBoolean(false);
    private int pingIntervalSeconds = 10; // Default value
    
    /**
     * Start heartbeat after application is ready
     */
    @EventListener(ApplicationReadyEvent.class)
    @Async
    public void startHeartbeat() {
        if (isRunning.compareAndSet(false, true)) {
            log.info("[{}] [HealthHeartbeatPublisher] Starting health heartbeat publisher", serviceName);
            
            // Load ping interval from Settings Service (self)
            try {
                String intervalStr = settingsService.getValue("HEALTH_PING_INTERVAL_SECONDS", "10");
                pingIntervalSeconds = Integer.parseInt(intervalStr);
                log.info("[{}] [HealthHeartbeatPublisher] Health ping interval: {} seconds", serviceName, pingIntervalSeconds);
            } catch (Exception e) {
                log.warn("[{}] [HealthHeartbeatPublisher] Failed to load HEALTH_PING_INTERVAL_SECONDS, using default: 10s. Error: {}", 
                    serviceName, e.getMessage());
            }
            
            // Create scheduler (single thread, daemon)
            scheduler = Executors.newSingleThreadScheduledExecutor(r -> {
                Thread t = new Thread(r, "health-heartbeat-" + serviceName);
                t.setDaemon(true);
                return t;
            });
            
            // Start periodic heartbeat
            scheduler.scheduleAtFixedRate(
                this::publishHealthStatus,
                0, // Initial delay: start immediately
                pingIntervalSeconds, // Period
                TimeUnit.SECONDS
            );
            
            log.info("[{}] [HealthHeartbeatPublisher] Health heartbeat publisher started (interval: {}s)", serviceName, pingIntervalSeconds);
        }
    }
    
    /**
     * Publish health status to Kafka
     */
    private void publishHealthStatus() {
        // Skip if we've encountered a fatal error (topic doesn't exist)
        if (hasError.get()) {
            return;
        }
        
        try {
            HealthStatusDto healthStatus = HealthStatusDto.builder()
                .serviceName(serviceName)
                .status("UP")
                .timestamp(LocalDateTime.now())
                .version(serviceVersion)
                .metadata(buildMetadata())
                .build();
            
            @SuppressWarnings("null")
            CompletableFuture<SendResult<String, Object>> future = 
                kafkaTemplate.send(TOPIC_HEALTH_STATUS, serviceName, healthStatus);
            
            future.whenComplete((result, ex) -> {
                if (ex == null) {
                    log.debug("[{}] [HealthHeartbeatPublisher] Health status published successfully", serviceName);
                } else {
                    log.error("[{}] [HealthHeartbeatPublisher] Failed to publish health status: {}", serviceName, ex.getMessage());
                    
                    // Check if error is due to topic not existing
                    if (ex.getCause() instanceof org.apache.kafka.common.errors.UnknownTopicOrPartitionException) {
                        log.error("[{}] [HealthHeartbeatPublisher] Topic '{}' does not exist. Stopping heartbeat.", serviceName, TOPIC_HEALTH_STATUS);
                        hasError.set(true);
                        stopHeartbeat();
                    }
                }
            });
            
        } catch (Exception e) {
            log.error("[{}] [HealthHeartbeatPublisher] Error publishing health status", serviceName, e);
        }
    }
    
    /**
     * Build metadata map for health status
     */
    private Map<String, Object> buildMetadata() {
        Map<String, Object> metadata = new HashMap<>();
        metadata.put("hostname", getHostname());
        metadata.put("pid", getProcessId());
        metadata.put("timestamp_millis", System.currentTimeMillis());
        return metadata;
    }
    
    private String getHostname() {
        try {
            return java.net.InetAddress.getLocalHost().getHostName();
        } catch (Exception e) {
            return "unknown";
        }
    }
    
    private String getProcessId() {
        return String.valueOf(ProcessHandle.current().pid());
    }
    
    /**
     * Stop heartbeat (called on error or shutdown)
     */
    private void stopHeartbeat() {
        if (scheduler != null && !scheduler.isShutdown()) {
            scheduler.shutdown();
            try {
                if (!scheduler.awaitTermination(5, TimeUnit.SECONDS)) {
                    scheduler.shutdownNow();
                }
            } catch (InterruptedException e) {
                scheduler.shutdownNow();
                Thread.currentThread().interrupt();
            }
        }
        isRunning.set(false);
    }
    
    /**
     * Shutdown hook
     */
    @PreDestroy
    public void destroy() {
        log.info("[{}] [HealthHeartbeatPublisher] Shutting down health heartbeat publisher", serviceName);
        stopHeartbeat();
    }
    
    /**
     * DTO for health status message
     */
    @Data
    @Builder
    public static class HealthStatusDto {
        private String serviceName;
        private String status;
        
        @JsonFormat(pattern = "yyyy-MM-dd'T'HH:mm:ss")
        private LocalDateTime timestamp;
        
        private String version;
        private Map<String, Object> metadata;
    }
}
